src/bsoncxx/builder/list.hpp

```diff
@@ -29,148 +30,119 @@ namespace builder {
-        std::stringstream err_msg{"cannot construct document"};
-        bool valid_document = false;
-        if (type_deduction || !is_array) {
-            valid_document = [&] {
-                if (init.size() % 2 != 0) {
-                    err_msg << " : must be list of key-value pairs";
-                    return false;
-                }
-                for (size_t i = 0; i < init.size(); i += 2) {
-                    auto t = (begin(init) + i)->val.view().type();
-                    if (t != type::k_utf8) {
-                        err_msg << " : all keys must be string type. ";
-                        err_msg << "Found type=" << to_string(t);
-                        return false;
-                    }
-                }
-                return true;
-            }();
-        }
-
-        if (valid_document) {
-            core _core{false};
-            for (size_t i = 0; i < init.size(); i += 2) {
-                _core.key_owned(std::string((begin(init) + i)->val.view().get_string().value));
-                _core.append((begin(init) + i + 1)->val);
-            }
-            val = bson_value::value(_core.extract_document());
-        } else if (type_deduction || is_array) {
-            core _core{true};
-            for (auto&& ele : init)
-                _core.append(ele.val);
-            val = bson_value::value(_core.extract_array());
-        } else {
-            throw bsoncxx::exception{error_code::k_unmatched_key_in_builder, err_msg.str()};
-        }
-    }
+    bool _is_document{false};
+    core _core{false};
+    bson_value::value _value{nullptr};
 };
 
 ///
-/// A JSON-like builder for creating documents.
+/// A JSON-like builder for creating BSON arrays.
 ///
-class document : public list {
-    using initializer_list_t = std::initializer_list<list>;
+class array {
+    using value_type = array;
+    using initializer_list_t = std::initializer_list<value_type>;
 
    public:
     ///
-    /// Creates an empty document.
+    /// Creates an empty array.
     ///
-    document() : list({}, false, false){};
+    array() : array({}){};
 
     ///
-    /// Creates a BSON document.
-    ///
-    /// @param init
-    ///     the initializer list used to construct the BSON document
-    ///
-    /// @see bsoncxx::builder::list
-    /// @see bsoncxx::builder::array
+    /// Creates an array from a single value of type T.
     ///
-    document(initializer_list_t init) : list(init, false, false) {}
-};
-
-///
-/// A JSON-like builder for creating arrays.
-///
-class array : public list {
-    using initializer_list_t = std::initializer_list<list>;
-
-   public:
+    /// @warning T must be a BSON type, i.e., implicitly convertible to a
+    /// bsoncxx::types::bson_value::value.
     ///
-    /// Creates an empty array.
+    /// @see bsoncxx::types::bson_value::value.
     ///
-    array() : list({}, false, true){};
+    template <typename T>
```

Comment:
Kind of the same comment here as on the equivalent document builder constructor: I don't think this adds a lot of value over the list constructor, and I think it would be simplest to omit it from the public API for this class (maybe we need it internally for some reason?)

Reply:
Removed.